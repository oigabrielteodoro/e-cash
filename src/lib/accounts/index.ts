export const agencyNumberWithoutDigitMask = '9999'
export const accountNumberWithDigitMask = '999999-9'
