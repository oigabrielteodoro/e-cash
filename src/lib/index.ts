export * from './matcher'
export * from './router'
export * from './formatters'
export * from './transformers'
export * from './input'
export * from './sidebar'
export * from './accounts'
export * from './masker'
export * from './text'
export * from './id'
export * from './email'
export * from './password'
export * from './url'
