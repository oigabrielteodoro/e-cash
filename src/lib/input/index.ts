export * from './useInput'
export * from './getValidationErrors'
