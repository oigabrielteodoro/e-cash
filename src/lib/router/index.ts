export * from './Router'

export const SIGN_IN = '/'
export const DASHBOARD = '/dashboard'
