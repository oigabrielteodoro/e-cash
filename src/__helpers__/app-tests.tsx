import React, { useEffect, ReactElement, ReactNode } from 'react'
import { render as rtlRender } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import {
  DefaultOptions,
  QueryCache,
  QueryClient,
  QueryClientProvider,
} from 'react-query'
import { MemoryRouter, Routes, Route } from 'react-router-dom'
import { generateId } from 'lib'
import nock from 'nock'

import { pipe } from 'fp-ts/function'
import { TaskEither, map, mapLeft } from 'fp-ts/TaskEither'

import { AppLayout, NotificationContainer } from 'ui'
import { clearToken, queryConfigDefault, setToken } from 'client'

type Options = {
  initialRoute?: string
  initialRoutePath?: string
  authenticated?: boolean
  queryClient?: QueryClient
  routePaths?: string[]
}

type Props = {
  children?: ReactNode
  client?: QueryClient
}

type Callback = (result: unknown) => unknown
type MapAll = (
  fn: Callback,
) => (data: TaskEither<unknown, unknown>) => TaskEither<unknown, unknown>

const queryCache = new QueryCache()
const defaultQueryClient = new QueryClient({
  defaultOptions: queryConfigDefault as DefaultOptions,
  queryCache: queryCache,
})

export function ReactQueryWrapper({
  children,
  client = defaultQueryClient,
}: Props) {
  return <QueryClientProvider client={client}>{children}</QueryClientProvider>
}

export function ReactRouterWrapper({
  initialRoute = '/',
  children,
  routePaths,
  authenticated = true,
}: Options & Pick<Props, 'children'>) {
  function MockedPage({ title }: { title: string }) {
    if (authenticated) {
      return (
        <AppLayout>
          <AppLayout.Content>Page {title}</AppLayout.Content>
        </AppLayout>
      )
    }

    return <h1>Page {title}</h1>
  }

  return (
    <MemoryRouter initialEntries={[initialRoute]}>
      <Routes>
        <Route path={initialRoute} element={children} />
        {routePaths?.map((routePath) => (
          <Route
            key={routePath}
            path={routePath}
            element={<MockedPage title={routePath} />}
          />
        ))}
      </Routes>
    </MemoryRouter>
  )
}

export function ReactHookWrapper({
  initialRoute = '/',
  children,
  routePaths,
  authenticated = true,
  client = defaultQueryClient,
}: Options & Props) {
  return (
    <ReactQueryWrapper client={client}>
      <ReactRouterWrapper
        initialRoute={initialRoute}
        routePaths={routePaths}
        authenticated={authenticated}
      >
        {children}
      </ReactRouterWrapper>
    </ReactQueryWrapper>
  )
}

function render(
  ui: ReactElement,
  {
    initialRoute = '/',
    queryClient = defaultQueryClient,
    authenticated = true,
    routePaths,
  }: Options = {},
) {
  function Wrapper({ children }: Props) {
    useEffect(() => {
      if (authenticated) {
        setToken(generateId())

        return () => {
          clearToken()
        }
      }
    }, [])

    return (
      <ReactQueryWrapper client={queryClient}>
        <ReactRouterWrapper
          initialRoute={initialRoute}
          routePaths={routePaths}
          authenticated={authenticated}
        >
          {children}
        </ReactRouterWrapper>

        <NotificationContainer />
      </ReactQueryWrapper>
    )
  }

  return {
    ...rtlRender(ui, {
      wrapper: Wrapper,
    }),
  }
}

export function unsafe<T>(value: unknown) {
  return value as T
}

export const mapAll: MapAll = (fn) => (data) => {
  return pipe(data, map(fn), mapLeft(fn))
}

export function getErrorMessage(errors: unknown): string {
  return Array.isArray(errors) ? errors[0].message : ''
}

export * from '@testing-library/react'
export {
  render,
  userEvent,
  nock,
  defaultQueryClient as queryClient,
  queryCache,
}
