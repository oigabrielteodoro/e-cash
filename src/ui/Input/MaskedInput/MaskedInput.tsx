import React, {
  useEffect,
  InputHTMLAttributes,
  ComponentType,
  ChangeEvent,
} from 'react'
import { AiOutlineExclamationCircle } from 'react-icons/ai'
import { useFormContext } from 'react-hook-form'

import type { IconBaseProps } from 'react-icons'

import { useInput, toMask, unMask, isResolvedMask } from 'lib'

import * as S from '../Input.styled'

export type MaskedInputProps = {
  mask: string[]
  name: string
  label: string
  error?: string
  variant?: 'primary' | 'secondary'
  icon: ComponentType<IconBaseProps>
} & InputHTMLAttributes<HTMLInputElement>

export function MaskedInput({
  mask,
  defaultValue,
  label,
  error,
  variant = 'primary',
  icon: Icon,
  name,
  id = name,
  onBlur,
  onFocus,
  onChange,
  ...rest
}: MaskedInputProps) {
  const { register, watch, trigger, setValue } = useFormContext()

  const value = watch(name)

  const {
    isFilled,
    isFocused,
    isErrored,
    errorMessage,
    handleOnBlur,
    handleOnFocus,
    setIsFilled,
  } = useInput({
    defaultValue,
    error,
    onBlur,
    onFocus,
  })

  useEffect(() => {
    register(name)
  }, [name, register])

  useEffect(() => {
    if (value) {
      setIsFilled(true)
    }
  }, [value, setIsFilled])

  const maskedValue = toMask(value, mask)

  function handleOnChange(event: ChangeEvent<HTMLInputElement>) {
    if (isResolvedMask(event.currentTarget.value, mask)) return

    setValue(name, unMask(event.currentTarget.value))
    trigger(name)

    onChange && onChange(event)
  }

  return (
    <S.Wrapper>
      <label htmlFor={id}>{label}</label>
      <S.Container
        variant={variant}
        isFilled={isFilled}
        isFocused={isFocused}
        isErrored={isErrored}
      >
        <input
          id={id}
          name={name}
          aria-label={name}
          defaultValue={defaultValue}
          onBlur={handleOnBlur}
          onFocus={handleOnFocus}
          onChange={handleOnChange}
          role='textbox'
          value={maskedValue}
          {...rest}
        />
        {Icon && <Icon size={22} />}
      </S.Container>
      {error && (
        <S.ErrorContainer>
          <AiOutlineExclamationCircle size={18} />
          <span>{errorMessage}</span>
        </S.ErrorContainer>
      )}
    </S.Wrapper>
  )
}
