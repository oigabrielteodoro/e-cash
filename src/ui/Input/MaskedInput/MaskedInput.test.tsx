import React from 'react'
import { useForm, FormProvider } from 'react-hook-form'
import { FiTag } from 'react-icons/fi'
import Faker from '@faker-js/faker'

import { render, screen, userEvent } from '__helpers__/app-tests'

import { toMask } from 'lib'
import { MaskedInput } from './MaskedInput'

type Props = {
  error?: string
  defaultValue?: string
}

const mask = ['9999']

function MockedComponent({ error, defaultValue }: Props) {
  const form = useForm({
    defaultValues: {
      mask: defaultValue,
    },
  })

  return (
    <FormProvider {...form}>
      <MaskedInput
        mask={mask}
        name='mask'
        label='Mask'
        icon={FiTag}
        error={error}
      />
    </FormProvider>
  )
}

describe('AmountInput', () => {
  it('should be able match snapshot', () => {
    const { container } = render(<MockedComponent />)

    expect(container).toMatchSnapshot()
  })

  it('should be able render correctly', () => {
    render(<MockedComponent />)

    expect(screen.getByRole('textbox')).toBeInTheDocument()
  })

  it('should be able render with initial values', async () => {
    const valueToType = Faker.datatype.number().toString()

    render(<MockedComponent defaultValue={valueToType} />)

    expect(await screen.findByRole('textbox')).toHaveValue(
      toMask(valueToType, mask),
    )
  })

  it('should be able type text correctly', async () => {
    const valueToType = Faker.datatype.number().toString()

    render(<MockedComponent />)

    const input = await screen.findByRole('textbox')

    expect(input).toBeInTheDocument()

    userEvent.type(input, valueToType)

    expect(await screen.findByRole('textbox')).toHaveValue(
      toMask(valueToType, mask),
    )
  })

  it('should be able render error message', async () => {
    render(<MockedComponent error='Value is required' />)

    expect(screen.getByRole('textbox')).toBeInTheDocument()
    expect(screen.getByText('Value is required')).toBeInTheDocument()
  })
})
