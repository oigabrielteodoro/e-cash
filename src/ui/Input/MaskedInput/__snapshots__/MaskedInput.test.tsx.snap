// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AmountInput should be able match snapshot 1`] = `
.c0 {
  border: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  position: relative;
}

.c0 label {
  font-size: 1.6rem;
  color: #a1a7ae;
  font-weight: 500;
  cursor: pointer;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: 0.1rem solid #a1a7ae;
  border-radius: 0.5rem;
  margin-top: 0.8rem;
  -webkit-transition: 300ms;
  transition: 300ms;
  width: 100%;
  background: #ffffff;
}

.c1 > svg,
.c1 .icon {
  margin-left: auto;
  margin-right: 2rem;
  color: #a1a7ae;
  -webkit-transition: 300ms;
  transition: 300ms;
}

.c1:hover {
  border: 0.1rem solid #3a67e1;
}

.c1:hover > svg,
.c1:hover .icon {
  color: #3a67e1;
}

.c1 input,
.c1 textarea {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  border: 0;
  background: transparent;
  padding: 1.8rem 2rem;
  color: #18171d;
  font-size: 1.6rem;
  border-radius: 0.5rem;
}

.c1 input::-webkit-input-placeholder,
.c1 textarea::-webkit-input-placeholder {
  color: #a1a7ae;
}

.c1 input::-moz-placeholder,
.c1 textarea::-moz-placeholder {
  color: #a1a7ae;
}

.c1 input:-ms-input-placeholder,
.c1 textarea:-ms-input-placeholder {
  color: #a1a7ae;
}

.c1 input::placeholder,
.c1 textarea::placeholder {
  color: #a1a7ae;
}

.c1 textarea {
  resize: none;
  padding: 0 2rem;
  margin: 2.4rem 0;
}

<div>
  <fieldset
    class="c0"
  >
    <label
      for="mask"
    >
      Mask
    </label>
    <div
      class="c1"
    >
      <input
        aria-label="mask"
        id="mask"
        name="mask"
        role="textbox"
        value=""
      />
      <svg
        fill="none"
        height="22"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="22"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M20.59 13.41l-7.17 7.17a2 2 0 0 1-2.83 0L2 12V2h10l8.59 8.59a2 2 0 0 1 0 2.82z"
        />
        <line
          x1="7"
          x2="7.01"
          y1="7"
          y2="7"
        />
      </svg>
    </div>
  </fieldset>
</div>
`;
